src/shared/resolve_or_timeout.ts

```diff
@@ -0,0 +1,22 @@
+import { timeout } from "./utils";
+
+interface TimeoutOrPromise<Type> {
+  timedOut: boolean;
+  result?: Type; // Populated if not timed out
+  promise?: Promise<Type>; // Populated if we did time out.
+}
+
```

Comment:
Please add a comment here explaining what this does. Comments right before the function declaration are shown in intellisense, it'll be helpful to readers who see this used elsewhere.